packages/@aws-cdk/aws-ecs-patterns/lib/base/application-multiple-target-groups-service-base.ts

```diff
@@ -0,0 +1,612 @@
+  /**
+   * Name of the listener.
+   */
+  readonly name: string;
+
+  /**
+   * The protocol for connections from clients to the load balancer.
+   * The load balancer port is determined from the protocol (port 80 for
+   * HTTP, port 443 for HTTPS).  A domain name and zone must be also be
+   * specified if using HTTPS.
+   *
+   * @default ApplicationProtocol.HTTP. If a certificate is specified, the protocol will be
+   * set by default to ApplicationProtocol.HTTPS.
+   */
+  readonly protocol?: ApplicationProtocol;
+
+  /**
+   * The port on which the listener listens for requests.
+   *
+   * @default - Determined from protocol if known.
+   */
+  readonly port?: number;
+
+  /**
+   * Certificate Manager certificate to associate with the load balancer.
+   * Setting this option will set the load balancer protocol to HTTPS.
+   *
+   * @default - No certificate associated with the load balancer, if using
+   * the HTTP protocol. For HTTPS, a DNS-validated certificate will be
+   * created for the load balancer's specified domain name.
+   */
+  readonly certificate?: ICertificate;
+}
+
+/**
+ * The base class for ApplicationMultipleTargetGroupsEc2Service and ApplicationMultipleTargetGroupsFargateService services.
+ */
+export abstract class ApplicationMultipleTargetGroupsServiceBase extends cdk.Construct {
+
+  /**
+   * The desired number of instantiations of the task definition to keep running on the service.
+   */
+  public readonly desiredCount: number;
+
+  /**
+   * The default Application Load Balancer for the service (first added load balancer).
+   */
+  public readonly loadBalancer: ApplicationLoadBalancer;
+
+  /**
+   * The default listener for the service (first added listener).
+   */
+  public readonly listener: ApplicationListener;
+
+  /**
+   * The cluster that hosts the service.
+   */
+  public readonly cluster: ICluster;
+
+  protected logDriver?: LogDriver;
+  protected listeners = new Array<ApplicationListener>();
+  protected targetGroups = new Array<ApplicationTargetGroup>();
+
+  private loadBalancers = new Array<ApplicationLoadBalancer>();
+
+  /**
+   * Constructs a new instance of the ApplicationMultipleTargetGroupsServiceBase class.
+   */
+  constructor(scope: cdk.Construct, id: string, props: ApplicationMultipleTargetGroupsServiceBaseProps = {}) {
+    super(scope, id);
+
+    this.validateInput(props);
+
+    this.cluster = props.cluster || this.getDefaultCluster(this, props.vpc);
+    this.desiredCount = props.desiredCount || 1;
+    if (props.taskImageOptions) {
+      this.logDriver = this.createLogDriver(props.taskImageOptions.enableLogging, props.taskImageOptions.logDriver);
+    }
+
+    if (props.loadBalancers) {
+      for (const lbProps of props.loadBalancers) {
+        const lb = this.createLoadBalancer(lbProps.name, lbProps.publicLoadBalancer);
+        this.loadBalancers.push(lb);
+        const protocolType = new Set<ApplicationProtocol>();
+        if (lbProps.listeners) {
```

Comment:
If `lbProps` does not have any listeners, does it make sense to fail early? I see we do a check in line 402, but not sure if that catches it for all load balancers specified in `props.loadbalancers`

Reply:
The intention is that if there's only one listener defined then it make sense to make that only listener as the default one. However, if there's no listener defined, then we fail. We actually allow them to have a lb without any listener if they want. Just you cannot have no listener at all. Otherwise there's no listener to attach the target group.

Reply:
However, since cloudformation will fail at deploy time if there's a LB without any listener. I guess we should prevent users from doing that.